Improve reporting of shader compiler errors

msteiger opened the pull request:
This PR refactors the GLSLShader class by splitting large methods into smaller ones. It also

* removes the highly fragile parsing of error logs and reports them as-is instead.
* removes the JOptionPane dialog and throws an exception instead.

Everything else remains the same: As usual, `AssetManagerImpl` logs the thrown exception and returns `null` instead. Callers can then decide to use a fall-back shader or throw an exception. Uncaugt exceptions will crash the application and trigger the CrashReporter.

I'm not overly happy with throwing `RuntimeException`. Anyone knows a better match?

Fixes #1649

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/87/
Hooray Jenkins reported success with all tests good!

Cervator commented:
Uh oh, this did something weird. Ran the game and got the following mostly black screen:

![terasology _ pre alpha_2015-05-19_23-13-01](https://cloud.githubusercontent.com/assets/1063833/7718351/2ae6d660-fe7d-11e4-95be-d95f46764589.png)

Nothing out of the ordinary in the logs.

Anybody else able to test?

Cervator commented:
Interesting. Retested just in case my computer is going crazy.

If I run without this branch and set all graphic settings to minimal/off then exit the game, checkout this branch, and run, the main menu loads fine.

Change graphic present back to "Insane!", exit game, start game, main menu messed up again.

@msteiger / anybody else: Try it on default settings and see if you can replicate?

emanuele3d commented:
Sorry I'm not finding the time to check on this. The black screen before the menu problem sounded familiar, but I seem to remember it occurred when shaders where initialized very early on startup, before the menu. If I remember correctly one of my PRs moved all shaders initialization to the initialization of the WorldRenderer, well away from the Menu state. So, it must be something else?

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/101/
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/102/
Hooray Jenkins reported success with all tests good!

msteiger commented:
I updated the PR: 

* fixed a bug that caused the black screen
* fixed all merge conflicts
* added a unit test (disabled on headless environments)

emanuele3d commented on engine-tests/src/main/java/org/terasology/DisplayEnvironment.java:
```diff
@@ -83,19 +85,28 @@ protected void setupDisplay() {
     protected void setupAssetManager() {
-        AssetManager assetManager = new AssetManagerImpl(context.get(ModuleManager.class).getEnvironment());
+        NullAudioManager audioManager = new NullAudioManager();
+        context.put(AudioManager.class, audioManager);
```
Can these two lines be moved to the end of the method, just before the two setAssetFactory calls are made?

emanuele3d commented on engine-tests/src/main/java/org/terasology/rendering/GLSLShaderTest.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    @Test
+    public void testOk() {
```
I have very little experience with unit testing and perhaps these are typical test names. If the aren't however, I'd rename testOk() and testBroken() to testConstructorWithValidShader() and testConstructorWithFaultyShader().

Also, is it worth testing what happens if a non AssetType.SHADER uri is provided?

emanuele3d commented on engine/src/main/java/org/terasology/rendering/opengl/GLSLShader.java:
```diff
@@ -279,25 +274,36 @@ private void updateAvailableFeatures() {
-        compileShader(GL20.GL_VERTEX_SHADER, features);
-    }
-
     private void compileAllShaderPermutations() {
```
Can I beg for a javadoc or a block comment for this one of all methods?

I can see what it does, but it'd be good to have an explanation of what "features" are. Also, am I right in thinking that this method just makes sure that all computations can be correctly compiled but then discards all compiled shaders? It'd be good to explain that too in the javadoc.

emanuele3d replied:
I looked into the code a bit better and I can see it holds two maps, one for vertex the other for fragment shaders, using a featureHash as key and the shaderId returned by glCreateShader() as value. So, the guess in the previous comment of mine is incorrect. Still, it'd be good to explain here or even in the javadoc for public method recompile() what this is doing.

But perhaps this is all out of the scope of this PR. I feel for example that an instance of this class is not, in fact, a shader in the opengl/computer graphic sense of the word, but the manager for a number of variants of the same shader. I also wonder if it's worth to keep all those variants in GPU memory compared to recompiling a shader on the fly when a feature is activated/deactivated. But as I said, this is probably out of the scope of this particular PR.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/opengl/GLSLShader.java:
```diff
@@ -279,25 +274,36 @@ private void updateAvailableFeatures() {
         }
-        logger.debug("Compiled {} permutations for {}.", counter, getURI());
+
+        Set<Set<ShaderProgramFeature>> allPermutations = Sets.powerSet(availableFeatures);
```
I think here you could add the Collections.emptySet() to the allPermutions set. This would allow you to remove the whole block above, as the empty set would be handled by the loop below. You would however need to handle the empty set as a special case for the error string given to the exception.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/opengl/GLSLShader.java:
```diff
@@ -279,25 +274,36 @@ private void updateAvailableFeatures() {
+                throw new RuntimeException(String.format("Shader '%s' failed to compile for feature set '%s'",
+                        getURI(), permutation));
+            }
+        }
```
A couple of issues here:
1) can we add the log info to the error message, like it is done for the empty set above?
2) right now as soon as a shader fails an exception is thrown and the loop stops. How about continuing for all permutations and throwing an exception only at the end, if at least one shader failed to compile? In this scenario, I don't know what's more appropriate: use logger.error() to send compilation errors to the log as soon as they are generated, with the exception being passed a static string, or simply add each error to a string and pass that to exception.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/opengl/GLSLShader.java:
```diff
@@ -322,85 +328,68 @@ private void compileShader(int type, Set<ShaderProgramFeature> features) {
             shader.append(shaderProgramBase.getVertexProgram());
         }
 
+        return shader.toString();
```
I know you didn't touch that bit, but couldn't the two if blocks above be conflated into one?

emanuele3d commented on engine/src/main/java/org/terasology/rendering/opengl/GLSLShader.java:
```diff
@@ -322,85 +328,68 @@ private void compileShader(int type, Set<ShaderProgramFeature> features) {
+        // Dump all final shader sources to the log directory
+        final String strippedTitle = getURI().toString().replace(":", "-");
+
+        String fname = debugShaderType.toLowerCase() + "_" + strippedTitle + "_" + featureHash + ".glsl";
```
Can you add a comment providing an example of what the filename would be like? I can sort of imagine it but it'd be good to have an example.

emanuele3d commented on engine/src/main/java/org/terasology/rendering/opengl/GLSLShader.java:
```diff
@@ -322,85 +328,68 @@ private void compileShader(int type, Set<ShaderProgramFeature> features) {
-            try {
-                Pattern p = Pattern.compile("-?\\d+");
-                Matcher m = p.matcher(error.toString());
+        Config config = CoreRegistry.get(Config.class);
```
The renderingConfig and the renderingDebugConfig are obtained and reobtained from the CoreRegistry throughout the class. Can we get them only once, into a couple of variables and use those?

emanuele3d commented on engine/src/main/java/org/terasology/rendering/opengl/GLSLShader.java:
```diff
@@ -279,25 +274,36 @@ private void updateAvailableFeatures() {
 
-    private void compileShader(int type, Set<ShaderProgramFeature> features) {
-        int shaderId = GL20.glCreateShader(type);
+    private String createShader(int type, Set<ShaderProgramFeature> features) {
```
I'm thinking this could be better called buildShader or assembleShader, also to differentiate it more from glCreateShader.

Cervator commented:
Interesting observation: new forum user reported in with a screenshot exactly like the one I got (but I only got the black screen actually captured from above). Weird part being said user gets it *now* while I only got it with an earlier version of this PR.

http://forum.terasology.org/threads/black-screen-major-bug-logs-and-screenshot.1316

There is also a weird delta time logged (in the menu log) that matches an older similar black (or dark?) screen issue.

As for me testing this I don't get the same situation with the newer commit. I do crash if I try to switch to the "Uber" graphics setting template. In this case I had started a normal game, changed to Minimal in-game, returned to main menu, started a new game, back to main menu, switch to uber, crash.

```
01:26:51.562 [Saving-0] INFO  o.t.p.internal.SaveTransaction - Save game finished
01:26:51.562 [main] WARN  org.reflections.Reflections - given scan urls are empty. set urls in the configuration
01:26:52.466 [main] WARN  o.t.engine.internal.TimeBase - Delta too great (2096), capping to 1000
01:26:57.735 [main] INFO  o.t.r.n.l.m.v.VideoSettingsScreen - Video Settings: {"pixelFormat":24,"windowPosX":365,"windowPosY":103,"windowWidth":1152,"windowHeight":720,"fullscreen":false,"viewDistance":"moderate","flickeringLight":true,"animateGrass":true,"animateWater":false,"fieldOfView":90.0,"cameraBobbing":false,"renderPlacingBox":true,"blurIntensity":2,"reflectiveWater":false,"vignette":true,"motionBlur":true,"ssao":true,"filmGrain":true,"outline":false,"lightShafts":true,"eyeAdaptation":true,"bloom":true,"dynamicShadows":false,"oculusVrSupport":false,"maxTextureAtlasResolution":4096,"maxChunksUsedForShadowMapping":1024,"shadowMapResolution":1024,"normalMapping":false,"parallaxMapping":false,"dynamicShadowsPcfFiltering":false,"cloudShadows":true,"renderNearest":true,"particleEffectLimit":10,"frameLimit":60,"meshLimit":400,"inscattering":true,"localReflections":false,"vSync":false,"clampLighting":false,"fboScale":100,"dumpShaders":false,"screenshotSize":"normal size","screenshotFormat":"jpg","cameraSettings":{"cameraSetting":"normal"},"debug":{"enabled":false,"stage":"opaque_color","firstPersonElementsHidden":false,"hudHidden":false,"wireframe":false,"renderChunkBoundingBoxes":false,"renderSkeletons":false}}
01:26:58.325 [main] ERROR o.terasology.engine.TerasologyEngine - Uncaught exception, attempting clean game shutdown
java.lang.NullPointerException: null
	at org.terasology.rendering.shader.ShaderParametersSSAO.applyParameters(ShaderParametersSSAO.java:95) ~[classes/:na]
	at org.terasology.rendering.opengl.GLSLMaterial.enable(GLSLMaterial.java:105) ~[classes/:na]
	at org.terasology.rendering.opengl.GLSLMaterial.restoreStateAfterUniformsSet(GLSLMaterial.java:601) ~[classes/:na]
	at org.terasology.rendering.opengl.GLSLMaterial.setFloat3(GLSLMaterial.java:392) ~[classes/:na]
	at org.terasology.rendering.assets.material.BaseMaterial.setFloat3(BaseMaterial.java:131) ~[classes/:na]
	at org.terasology.rendering.shader.ShaderParametersSSAO.initialParameters(ShaderParametersSSAO.java:88) ~[classes/:na]
	at org.terasology.rendering.opengl.GLSLMaterial.recompile(GLSLMaterial.java:138) ~[classes/:na]
	at org.terasology.rendering.ShaderManagerLwjgl.recompileAllShaders(ShaderManagerLwjgl.java:164) ~[classes/:na]
	at org.terasology.rendering.nui.layers.mainMenu.videoSettings.VideoSettingsScreen.onClosed(VideoSettingsScreen.java:237) ~[classes/:na]
	at org.terasology.rendering.nui.internal.NUIManagerInternal.popScreen(NUIManagerInternal.java:298) ~[classes/:na]
	at org.terasology.rendering.nui.layers.mainMenu.videoSettings.VideoSettingsScreen$7.onActivated(VideoSettingsScreen.java:229) ~[classes/:na]
	at org.terasology.rendering.nui.widgets.UIButton.activate(UIButton.java:138) ~[classes/:na]
	at org.terasology.rendering.nui.widgets.UIButton.access$100(UIButton.java:40) ~[classes/:na]
	at org.terasology.rendering.nui.widgets.UIButton$1.onMouseRelease(UIButton.java:81) ~[classes/:na]
	at org.terasology.rendering.nui.internal.CanvasImpl.processMouseRelease(CanvasImpl.java:232) ~[classes/:na]
	at org.terasology.rendering.nui.internal.NUIManagerInternal.mouseButtonEvent(NUIManagerInternal.java:559) ~[classes/:na]
	at org.terasology.rendering.nui.internal.NUIManagerInternalMethodAccess.invoke(Unknown Source) ~[reflectasm-1.11.0.jar:na]
	at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:506) ~[classes/:na]
	at org.terasology.entitySystem.event.internal.EventSystemImpl.sendConsumableEvent(EventSystemImpl.java:280) ~[classes/:na]
	at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:259) ~[classes/:na]
	at org.terasology.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:157) ~[classes/:na]
	at org.terasology.input.InputSystem.sendMouseEvent(InputSystem.java:393) ~[classes/:na]
	at org.terasology.input.InputSystem.processMouseInput(InputSystem.java:248) ~[classes/:na]
	at org.terasology.input.InputSystem.update(InputSystem.java:204) ~[classes/:na]
	at org.terasology.engine.modes.StateMainMenu.handleInput(StateMainMenu.java:140) ~[classes/:na]
	at org.terasology.engine.subsystem.lwjgl.LwjglInput.postUpdate(LwjglInput.java:66) ~[classes/:na]
	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:486) ~[classes/:na]
	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:397) ~[classes/:na]
	at org.terasology.engine.Terasology.main(Terasology.java:151) [classes/:na]
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_45]
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_45]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_45]
	at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_45]
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140) [idea_rt.jar:na]
01:26:58.325 [main] INFO  o.terasology.engine.TerasologyEngine - Shutting down Terasology...
java.lang.NullPointerException
	at org.terasology.rendering.shader.ShaderParametersSSAO.applyParameters(ShaderParametersSSAO.java:95)
	at org.terasology.rendering.opengl.GLSLMaterial.enable(GLSLMaterial.java:105)
	at org.terasology.rendering.opengl.GLSLMaterial.restoreStateAfterUniformsSet(GLSLMaterial.java:601)
	at org.terasology.rendering.opengl.GLSLMaterial.setFloat3(GLSLMaterial.java:392)
	at org.terasology.rendering.assets.material.BaseMaterial.setFloat3(BaseMaterial.java:131)
	at org.terasology.rendering.shader.ShaderParametersSSAO.initialParameters(ShaderParametersSSAO.java:88)
	at org.terasology.rendering.opengl.GLSLMaterial.recompile(GLSLMaterial.java:138)
	at org.terasology.rendering.ShaderManagerLwjgl.recompileAllShaders(ShaderManagerLwjgl.java:164)
	at org.terasology.rendering.nui.layers.mainMenu.videoSettings.VideoSettingsScreen.onClosed(VideoSettingsScreen.java:237)
	at org.terasology.rendering.nui.internal.NUIManagerInternal.popScreen(NUIManagerInternal.java:298)
	at org.terasology.rendering.nui.layers.mainMenu.videoSettings.VideoSettingsScreen$7.onActivated(VideoSettingsScreen.java:229)
	at org.terasology.rendering.nui.widgets.UIButton.activate(UIButton.java:138)
	at org.terasology.rendering.nui.widgets.UIButton.access$100(UIButton.java:40)
	at org.terasology.rendering.nui.widgets.UIButton$1.onMouseRelease(UIButton.java:81)
	at org.terasology.rendering.nui.internal.CanvasImpl.processMouseRelease(CanvasImpl.java:232)
	at org.terasology.rendering.nui.internal.NUIManagerInternal.mouseButtonEvent(NUIManagerInternal.java:559)
	at org.terasology.rendering.nui.internal.NUIManagerInternalMethodAccess.invoke(Unknown Source)
	at org.terasology.entitySystem.event.internal.EventSystemImpl$ByteCodeEventHandlerInfo.invoke(EventSystemImpl.java:506)
	at org.terasology.entitySystem.event.internal.EventSystemImpl.sendConsumableEvent(EventSystemImpl.java:280)
	at org.terasology.entitySystem.event.internal.EventSystemImpl.send(EventSystemImpl.java:259)
	at org.terasology.entitySystem.entity.internal.BaseEntityRef.send(BaseEntityRef.java:157)
	at org.terasology.input.InputSystem.sendMouseEvent(InputSystem.java:393)
	at org.terasology.input.InputSystem.processMouseInput(InputSystem.java:248)
	at org.terasology.input.InputSystem.update(InputSystem.java:204)
	at org.terasology.engine.modes.StateMainMenu.handleInput(StateMainMenu.java:140)
	at org.terasology.engine.subsystem.lwjgl.LwjglInput.postUpdate(LwjglInput.java:66)
	at org.terasology.engine.TerasologyEngine.mainLoop(TerasologyEngine.java:486)
	at org.terasology.engine.TerasologyEngine.run(TerasologyEngine.java:397)
	at org.terasology.engine.Terasology.main(Terasology.java:151)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.intellij.rt.execution.application.AppMain.main(AppMain.java:140)
```

msteiger commented:
I think I addressed all of your concerns, @emanuele3d - thanks for reviewing.

And well spotted, @Cervator. The bug seems to be unrelated to this PR as it can be reproduced without the changes here. I opened a new issue for it.

Cervator commented:
Cool :-)

Should we merge this ahead of #1740 or wait? This PR has a slight bit of Asset interaction.

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/122/

**Build Log**
*last 5 lines*

 ```
[...truncated 2389 lines...]
	at hudson.model.Executor.run(Executor.java:240)
[ANALYSIS-COLLECTOR] Skipping publisher since build result is ABORTED
Notifying upstream projects of job completion
Setting status of 73abb8b0a731978a7975b455cd1e7cd9f119c0fa to FAILURE with url http://jenkins.terasology.org/job/TerasologyPRs/122/ and message: Merged build finished.
```

Uh oh, something went wrong with the build. Need to check on that

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/123/

**Build Log**
*last 5 lines*

 ```
[...truncated 232 lines...]
	at hudson.model.Executor.run(Executor.java:240)
[ANALYSIS-COLLECTOR] Skipping publisher since build result is ABORTED
Notifying upstream projects of job completion
Setting status of 54762d03f608863c543f06777bc85cc3c66695bc to FAILURE with url http://jenkins.terasology.org/job/TerasologyPRs/123/ and message: Merged build finished.
```

Uh oh, something went wrong with the build. Need to check on that

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/126/
Hooray Jenkins reported success with all tests good!

msteiger commented:
Thank you for the review. I fixed the documentation.

I think that it can be merged before #1740 since `GLSLShader` is not touched. `DisplayEnvironment` is on the deletion list there (probably because it was not used before), but I'd like to keep for testing graphics-related stuff.

msteiger commented:
Is there anything still unclear or unresolved here? If not, then it can be merged, I think.

Cervator commented:
Will be happy to try to merge this tomorrow after the big one, just in case from the slight bit of asset interaction it has :-)

emanuele3d commented:
I confirm this is good to go also for me. Good job Martin!

Cervator commented:
Annnnd more need to resolve conflicts / rebase :-)

Since this has been waiting a while I went to see what was being problematic and the main thing seems to be `DisplayEnvironment` having gone missing. Not sure where it got moved / what it was renamed to, or if something made it obsolete.

@immortius do you recall off-hand what the fate was for that in the big merge, if it was involved? It seems to go poof in that period yet #1740 doesn't have it in the file list. Quirk from merges/rebase maybe.

immortius commented:
DisplayEnvironment wasn't used anywhere, so I saw no reason to keep it.

Cervator commented:
Ah, so the new test introduced in this PR was based on it but then it was deleted.

@msteiger does it make sense to use one of the other environments or do we have a use case for the DisplayEnvironment now? Since it tests shaders it might make sense that it needs a display ..

GooeyHub commented:
Refer to this link for build results (access rights to CI server needed): 
http://jenkins.terasology.org/job/TerasologyPRs/138/
Hooray Jenkins reported success with all tests good!

Cervator commented:
Noticed rebase had been done, tested quickly, merged since we've gone over this plenty before :-)

Feel free to comment so it'll ping the PR :D
